### ISSUE TITLE
Invoke-HKCURegistrySettingsForAllUsers - Failed to unload

### ISSUE THREAD

**Issue by thomashesse** [PSAppDeployToolkit/PSAppDeployToolkit#66]
Env
* Windows 7 x64
* SCCM 2012 R2
* PSAppDeployToolkit 3.6.5

Cmdlet / Code
```
[scriptblock]$HKCURegistrySettings = {
			Set-RegistryKey -Key 'HKCU\Software\...\...' -Name '...' -Value '...' -Type String -SID $UserProfile.SID
		}
		Invoke-HKCURegistrySettingsForAllUsers -RegistrySettings $HKCURegistrySettings
```
Error
```
[Post-Installation] :: Load the User [...] registry hive in path [HKEY_USERS\...].	Invoke-HKCURegistrySettingsForAllUsers
[Post-Installation] :: Execute ScriptBlock to modify HKCU registry settings for all users.	Invoke-HKCURegistrySettingsForAllUsers
[Post-Installation] :: Return fully qualified registry key path [Registry::HKEY_USERS\...].	Convert-RegistryPath
[Post-Installation] :: Update registry key value: [Registry::HKEY_USERS\...] [... = ...].	Set-RegistryKey
[Post-Installation] :: Performing manual garbage collection to ensure successful unloading of registry hive.	Invoke-HKCURegistrySettingsForAllUsers
[Post-Installation] :: Unload the User [...] registry hive in path [...].	Invoke-HKCURegistrySettingsForAllUsers
[Post-Installation] :: Failed to unload the registry hive for User [...] with SID [...]. 
Error Record:
-------------

Message        : REG.exe failed with exit code [1] and result [].
InnerException : 

FullyQualifiedErrorId : REG.exe failed with exit code [1] and result [].
ScriptStackTrace      : bei Invoke-HKCURegistrySettingsForAllUsers<Process>, C:\Temp\...PSADT 
                        v3.6.5\AppDeployToolkit\AppDeployToolkitMain.ps1: Zeile 3608
                        bei <ScriptBlock>, C:\Temp\...PSADT v3.6.5\Deploy-Application.ps1: Zeile 
                        137
                        bei <ScriptBlock>, <Keine Datei>: Zeile 1
                        bei <ScriptBlock>, <Keine Datei>: Zeile 1

PositionMessage : In C:\Temp\...PSADT v3.6.5\AppDeployToolkit\AppDeployToolkitMain.ps1:3608 
                  Zeichen:41
                  +                         If ($global:LastExitCode -ne 0) { Throw "REG.exe failed with exit code [$( ...
                  +    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by mmashwani** [comment]
Does this happen consistently...every time you run the script?

**Comment by thomashesse** [comment]
The error occured 5 times in 5 tests. Always (failed to unload) concerned different users.

**Comment by mmashwani** [comment]
Did it fail to unload a specific registry hive or all of them?

**Comment by thomashesse** [comment]
Some failed, not all. It failed at one Domain User, one local user and the default user.

**Comment by mmashwani** [comment]
Can you try using handle.exe from sysinternals to see if there is anything that has a handle on the .dat file that represents the registry for a specific user?

**Comment by thomashesse** [comment]
I see no entry for the specific users.Only entries on the ntuser.dat for the logged in user and LocalService.

**Comment by mmashwani** [comment]
Hmm, I'm not sure what to say then. I will tag this post with the "help wanted" tag. If you or anybody else can provide further information on this situation so that we can replicate the issue or understand it better, then we can move forward with some sort of fix.

**Comment by DennisVH** [comment]
I has similar issues unloading hives in a self-written powershell function.
I solved it by doing a [gc]::collect() after performing the unload when the unload result was not equal to 0, wait 5 seconds and then rerun the unload.

$result = Execute-Process -FilePath "$env:windir\System32\reg.exe" -Arguments "UNLOAD $TempHive" -PassThru -CreateNoWindow -WindowStyle Hidden
if ($result.ExitCode -ne 0)
{
	[gc]::collect()
	Start-Sleep -s 5
	Start-Process -FilePath "$env:windir\System32\reg.exe" -ArgumentList "UNLOAD $TempHive" -Wait
}

I tried to recreate the issue, but was not able to.
